xir/parser.py

```diff
@@ -174,70 +173,28 @@ def expr(self, args):
     # definitions and statements
     ##############################
 
-    def gate_def(self, args):
+    @v_args(inline=True)
+    def gate_def(self, name, params_list, wires, *stmts):
         """Gate definition. Starts with keyword 'gate'. Adds gate to program."""
-        name = args.pop(0)
-        wires = ()
-        params = []
-        stmts = []
 
-        max_wire = 0
         has_declared_wires = False
-        for i, arg in enumerate(args):
-            if is_param(arg):
-                params = arg[1]
-            elif is_wire(arg):
-                has_declared_wires = True
-                wires = arg[1]
-            elif isinstance(arg, Statement):
-                if has_declared_wires:
-                    stmts = args[i:]
-                    break
-                stmts.append(arg)
-
-                int_wires = [w for w in arg.wires if isinstance(w, int)]
-                if int_wires and max(int_wires) > max_wire:
-                    max_wire = max(int_wires)
+        if wires is not None and len(wires) > 0:
+            has_declared_wires = True
 
-            if not has_declared_wires:
-                wires = tuple(range(max_wire + 1))
-            else:
-                # remove duplicate wires while maintaining order
-                wires = tuple(dict.fromkeys(wires))
-
-        self._program.add_gate(name, params, wires, stmts)
-
-    def obs_def(self, args):
-        """Observable definition. Starts with keyword 'obs'. Adds observable to program."""
-        name = args.pop(0)
-        wires = ()
-        params = []
-        stmts = []
 
-        max_wire = 0
-        has_declared_wires = False
-        for i, arg in enumerate(args):
-            if is_param(arg):
-                params = arg[1]
-            elif is_wire(arg):
-                wires = arg[1]
-                has_declared_wires = True
-            elif isinstance(arg, ObservableStmt):
-                if has_declared_wires:
-                    stmts = args[i:]
-                    break
-                stmts.append(arg)
-
-                int_wires = [w for w in arg.wires if isinstance(w, int)]
+        if not has_declared_wires:
+            max_wire = 0
+            for stmt in stmts:
+                int_wires = [w for w in stmt.wires if isinstance(w, int)]
                 if int_wires and max(int_wires) > max_wire:
                     max_wire = max(int_wires)
 
-            if not has_declared_wires:
-                wires = tuple(range(max_wire + 1))
-            else:
-                wires = tuple(dict.fromkeys(wires))
+            wires = tuple(range(max_wire + 1))
+        else:
+            # remove duplicate wires while maintaining order
+            wires = tuple(dict.fromkeys(wires))
 
-        self._program.add_observable(name, params, wires, stmts)
+        self._program.add_gate(name, [] if params_list is None else params_list[1], wires, stmts)
```

Comment:
Generally, it's cleaner to leave `x if y else z` out of long expressions:
```suggestion
        params =  [] if params_list is None else params_list[1]
        self._program.add_gate(name, params, wires, stmts)
```

Reply:
fixed